docs/process/experimental-flags.md

```diff
@@ -0,0 +1,98 @@
+# Dart SDK process for changes behind experimental flags
+
+## Problem statement
+
+The Dart SDK ships via a number of channels:
+
+- Via the [Dart SDK](https://www.dartlang.org/tools/sdk#install)
+- Via the [Flutter SDK](http://flutter.io)
+- Internally at Google via an internal channel
+
+Each of these channels use varying release calendars, and keeping these entirely
+aligned is not practical. Further, a number of developers interested in staying
+current with Dart changes, consume the Dart SDK via our [dev
+channel](https://github.com/dart-lang/sdk/wiki/Branches-and-releases). As a
+result, we should anticipate that any dev channel build has the potential to end
+up as a shipped SDK though some channel. And as a consequence of that, it is
+critical that we keep the quality of our dev channel high AND that we keep it
+consistent wrt. which features are complete and supported. At the same time, we
+need the ability to land incomplete features to facilitate a number of tasks:
+
+- Compose a feature that spans multiple components/areas
+- Automated testing prior to the full completion of the feature
+- Allow partner teams to get a preview of a future feature
+- Allow customers to get a preview of a future feature
+- Etc.
+
+## Solution
+
+To ensure completed & supported features can be differentiated from incomplete
+features in-progress, we will put all in-progress features behind a single set
+of flags. Changes to features behind these flags are not considered breaking
+(even if the feature behind the flag was in a stable SDK), and they are subject
+to removal at any time. For details about breaking changes, see the breaking
+change process.
+
+All new features that meet one of the following criteria must be developed
+behind a flag (when possible):
+
+- All breaking changes
+- All language changes
+
+Further, it is recommended to consider developing behind a flag when:
+
+- Landing larger, user-visible changes which will be in an intermediate state
+  over several weeks and perhaps even releases
+- Making changes with the potential to have significant negative performance
+  impact for several weeks and perhaps even across releases
+
+## Details
+
+### Flag format for CLI-based tools
+
+Flags consist of one or more words, combined using dashes, using all lower-case.
+The single source of truth of these flags shall be a single shared .dart file.
+The tools are expected to offer a framework for querying these flags so that the
+implementation of the tools can easily access new flags.
+
+The flags are passed to CLI-based tools using the `--enable-experiment` flag
+(short-form `-e`). Multiple flags can be passed by using multiple flags, or by
+passing several comma-separated flags. Examples:
+
+```
+dart --enable-experiment super-mixins
+dart -e super-mixins
+dart -e super-mixins,no-slow-checks,preview-dart3
+dart -e super-mixins -e no-slow-checks -e preview-dart3
```

Comment:
Is "no-slow-checks" saying "disable the experiment 'slow-checks'" or is it enabling the experiment "no-slow-checks"?

Relevant to that, other similar systems I've seen have a way from the shell (sometimes alongside --version output) to list the set of known experiments and to give the list of experiments currently enabled by default.  There should also be a flag to explicitly disable one or more experiments.  This will allow for cases where a feature is almost ready for release (e.g. the new Dart supermixins) to be enabled by default, but we can disable it by flag for certain customers running into trouble.

I would expect an experiment could go through several phases in its lifetime:

- Unavailable -- Tools don't know about it and error if it is referenced
- Available, disabled by default -- Use --enable-experiment to turn it on
- Available, enabled by default -- Use --disable-experiment to turn it off
- Available, with a warning if disabled manually -- Use --disable-experiment to turn it off, but it warns the user that the flag is going away soon
- Permanently available -- Using --disable-experiment to try to turn it off is not allowed.

Reply:
Yes, a flag starting with `no` would generally disable an experiment. For example, had we used the present scheme for the Dart 2 launch we would initially have had a `--enable-experiment dart-2-type-system` (the opt-in; similar to `--preview-dart-2`), and later a `--enable-experiment no-dart-2-type-system` (the opt-out; similar to `--no-preview-dart-2`.

Yes, this is a little less elegant than something like your suggestion, but opt-out flags are typically only used for really large scale changes, which are highly breaking, which we don't expect to be making again. Thus we don't plan on supporting a `--disable-experiment` like syntax.

Reply:
I disagree here. You only need to turn *on* language experiments. There will not be a time where it is still an experiement, and it is enabled by default. If it is turned on by default, the flag will no longer have any effect, and it cannot be disabled by users.

There should be no "no-" prefix at all on language experiments. Since the option can be passed more than once, that would require us to decide what `--enable-experiment=foo --enable-experiment=no-foo` means.

Ofcourse, this is only my opinion for *language expriments*. If platforms or tools want to use the same syntax for other flags (not specified in the same file as the language options), then nothing prevents that, and those flags may be negatable. So, at least it should be able to not be negatable for a particular flag.

Reply:
@mit-mit  As long as there is some way to explicitly disable an experiment, the specifics I'm not as tied to.

@lhrn  That may be technically true from a language definition point of view, but retaining the ability to disable a new thing (language or otherwise) for a short time after it becoming the default has utility value when attempting to keep bugs from preventing deployments.  Turning on an experiment/feature by default has a tendency to find all the scenarios and projects who somehow missed your 47 deprecation warnings, emails, and automated tests and what have you to try to detect them, yet they can still matter when it comes to blocking rollouts.  It is far easier to have that option from the get-go then to wish you had it when you're in a sticky situation and a roll is blocked.

Reply:
@lhrn  If you like, remove phase 3 from my suggested experiment progression and minimize the amount of time a new language experiment can be disabled once it is active by default.  It's entirely reasonable not to leave them able to be switched off indefinitely.

Reply:
I think @jcollins-g point here is a good one.  We may have a better time rolling things out if we don't have to do a global rollback each time we discover one customer who isn't prepared for the change.

Reply:
If we link features to language versions, then you should be able to stay at an older language version for a while before you are forced to migrate. That works for both breaking and non-breaking changes. That is, if we have language versioning, we should not need *feature* versioning as well.

The moment a feature is enabled by default, it's fair game for implementations to start assuming it is there and make code that binds one feature tightly to another. Allowing opt-out of one particular feature may not be possible.
Or put differently: If we add a period where you can disable a feature again, we also delay the implementations' ability to incorporate the feature.

(Until we actually have language versioning, it is true that we may have to give some other way to avoid a new feature for unmigrated code).

Reply:
I'm trying to lock down a syntax for flags in this PR, not capture a comprehensive policy for how we might use that syntax. Hopefully we can lock down the syntax, whereas for the policy we will have to adjust as we learn. I'm going to resolve this comment thread based on that; if you disagree with the syntax, then please reactivate.

Reply:
I don't think we've resolved if we need syntax for disabling the feature have we?  

> If we link features to language versions, then you should be able to stay at an older language version for a while before you are forced to migrate. That works for both breaking and non-breaking changes. That is, if we have language versioning, we should not need _feature_ versioning as well.

I don't believe this.  I think these are essentially unrelated.  What is the process for rolling into internal?  If the process is:

- convert any code that needs changing, turning on  the flag
- enable by default in the sdk
- remove the flag internally

then maybe this works, since during the migration process if you break somebody you can selectively remove the "enable" flag for their target.  I'm still worried about the scenario that we:

- enable everywhere
- land enable by default in internal
- discover that a key customer is broken
- have to roll back instead of just disabling selectively